swupd/manifest.go

```diff
@@ -480,11 +487,52 @@ func (m *Manifest) linkPeersAndChange(oldManifest *Manifest, c config, minVersio
 func (m *Manifest) newDeleted(df *File) {
 	df.Version = m.Header.Version
 	df.Status = StatusDeleted
+	df.Type = TypeUnset
 	df.Modifier = ModifierUnset
 	// Add file to manifest
 	m.Files = append(m.Files, df)
 }
 
+func (f *File) renameFrom() bool {
+	return f.Rename == true && !f.Present()
+}
+
+func (f *File) renameTo() bool {
+	return f.Rename == true && f.Present()
+}
+
+// linkOldRenames links renames present in a manifest already
+// this should be done before renameDetection is run
+func (m *Manifest) linkOldRenames() {
+	for _, f1 := range m.Files {
```

Comment:
This is O(n^2) with large n. Make short lists of just things with an existing rename flag (O(n)) to reduce the amount of work and remove the need for the rename{From,To} functions.

Then sort both by hash, and do the linear pass over the hashes.

Possibly return the (maybe sorted) list so deprecateOldRename can use it to walk a much shorter list.

Reply:
yeah ok. I took out a comment at the top of this function that explained that although this was O(n^2), it didn't matter because we would be removing rename flags anyways soon.

It isn't difficult to make this efficient. I'll do so.

Reply:
I usually review the code as posted, not as I think the person intended to write it.

Reply:
@icarus-sparry I intended to write it the way I posted it, for the reason I gave above. I was only giving you context.